### PR TITLE
Fetch and display content from Java servlet using javascript

### LINKED ISSUE #6: Add fetch in JavaScript to be able to get data from the server
Add javascript that fetches data from the Java servlet

@billie-cat @f-camp

### PR DESCRIPTION
**Pull request by n-wach**
Step 2 of the adding comments tutorial.  Here I display the result from a `fetch` to the servlet at `/data`.

closes #6

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -34,3 +34,19 @@ function openProject() {
...
+/**
+ * Fetch and display a blurb from /data
+ */
+
+const BLURB_ENDPOINT = "/data";
+const BLURB_SELECTOR = "#blurb"
+
+async function showBlurb() {
+  let blurb = document.querySelector(BLURB_SELECTOR);
+  let request = await fetch(BLURB_ENDPOINT);
```
Comment (f-camp) [inline portfolio/src/main/webapp/script.js:48]: Prefer const here please.
https://google.github.io/styleguide/jsguide.html#features-use-const-and-let

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -20,13 +20,21 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns something to display on index.html. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  String[] blurbs = {
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:27]: private static final String [] blurbs = ...

File: portfolio/src/main/webapp/script.js (hunk 1 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -20,13 +20,21 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns "Hello Nathan". TODO: modify this file to handle comments data */
+/** Servlet that returns something to display on index.html. TODO: modify this file to handle comments data */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:23]: Nit: Someone reading this doesn't know what index.html is. Instead of linking can you describe where it is displayed. Eg. Nathan's personal site.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -40,9 +40,12 @@ for(let project of document.getElementsByClassName(PROJECT)) {
  * Fetch and display a blurb from /data
  */
 
+const BLURB_ENDPOINT = "/data";
+const BLURB_SELECTOR = "#blurb"
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:44]: nit: move these to the top of the file. Typically all const vars are together and at the top of the file (i.e. where OPEN) is.

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -40,9 +40,12 @@ for(let project of document.getElementsByClassName(PROJECT)) {
  * Fetch and display a blurb from /data
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:40]: This JsDoc comment should be above showBlurb (with no extra white spaces). 

Also, from the comment I am not sure what /data means. Can you describe that?

**Comment by billie-cat** [comment]
Can the title of this pull request be more descriptive? For example: Display the result from a fetch to the servlet at /data

**Comment by billie-cat** [comment]
> Can the title of this pull request be more descriptive? For example: Display the result from a fetch to the servlet at /data

for more context as to why I asked for this change: https://google.github.io/eng-practices/review/developer/cl-descriptions.html

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: I just updated comments and reorganized script.js--see latest commit

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Clarified this

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Done

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -14,23 +14,43 @@
...
-// This script is loaded with the `defer` tag, so we know the DOM is loaded and ready.
+// This script is loaded with the `defer` tag, so we know the DOM is loaded 
+// and ready.
 for(let project of document.getElementsByClassName(PROJECT)) {
   project.addEventListener("click", openProject);
 }
+
+
+/**
+ * Fetch and display a welcome message from the servlet at /data
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:47]: When you're writing reference documentation for a method, phrase the main method description in terms of what the method does (gets, lists, creates, searches), rather than what the developer would use it to do (get, list, create, search).

https://developers.google.com/style/reference-verbs

So here: I would write: Fetches and displays a welcome message obtained from the servlet...

Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:47]: I don't know much about web dev, what is /data? Is there a more descriptive name for it? This seems to be a variable.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -14,23 +14,43 @@
...
 const PROJECT = "project";
-
 // CSS class given to a project div to make it expand and reveal its content
 const OPEN = "open";
 
+const WELCOME_ENDPOINT = "/data";
+const WELCOME_SELECTOR = "#welcome"
+
+
+/**
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:29]: When you're writing reference documentation for a method, phrase the main method description in terms of what the method does (gets, lists, creates, searches), rather than what the developer would use it to do (get, list, create, search).

https://developers.google.com/style/reference-verbs

I would rephrase: Makes the project div clickable. When clicked it expands or closes based on ....

I would describe when it is expanded or when its closed so people do not need to look at the code to know what this method does.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -15,26 +15,31 @@
...
-/** Servlet that returns something to display on index.html. TODO: modify this file to handle comments data */
+/** 
+  * Servlet that returns a welcome message to display on Nathan's site. 
+  * TODO: modify this file to handle comments data 
+  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private static final String[] BLURBS = {
+  private static final ImmutableList<String> WELCOMES = ImmutableList.of(
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: Are you planning on adding more messages? Can this be only a String instead of a List?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -15,18 +15,31 @@
...
 import java.io.IOException;
+import com.google.common.collect.ImmutableList;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** 
+  * Servlet that returns a welcome message to display on Nathan's site.
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: nit: Servlet that returns a welcome message displayed on Nathan's site

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -15,18 +15,31 @@
...
+    "Hi, I'm Nathan.",
+    "Welcome!",
+    "Hello!"
+  );
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    String welcome = WELCOMES.get((int)(Math.random() * WELCOMES.size()));
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: ni:(int)(Math.random() * WELCOMES.size()
can be extracted into its own variable for clarity.

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:47]: I created the DataServlet java class, and told it to listen at the arbitrary endpoint "/data".  So if you go to [my website url]/data, the DataServlet .doGet() method is called and returns some data.  The choice of "/data" is arbitrary and I'll change it to "/comments" eventually (tutorials still refer to it as /data).  Would it make more sense to say `Fetch and display a welcome message from DataServlet at "/data"`?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: These are just hardcoded messages for testing.  The styleguide states that a String[] is not a valid constant (element reassignment), so I switched to using ImmutableList.  I'm returning a random one of these strings, so I can't condense it down to a single string.

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Fixed

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:29]: So this function doesn't actually make them clickable, it's just the event handler I add for click events.  Not sure what the tense should be.  I changed it a bit, please take a look at latest commit

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: Thanks will keep in mind. Not gonna change it here tho cause it's just for testing and already gone in the latest comments-step-3 branch

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: Thanks will keep in mind. Not gonna change it here tho cause it's just for testing and already gone in the latest comments-step-3 branch

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: Oh, I think I looked at an old change where there is only one. This makes sense.

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (billie-cat) [reply portfolio/src/main/webapp/script.js:29]: How about Handles click event for project div. Would that work?

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -14,23 +14,46 @@
...
+ * Event handler for project div click
+ * When clicked and closed, expands and scrolls the div into view
+ * When clicked and open, shrinks back to smaller size
+ */
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:33]: nit: Make these into proper sentences and no need to start a new line for each sentence.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -14,23 +14,46 @@
...
+// Add the openProject click event handler to every project div
+// This script is loaded with the `defer` tag, so we know the DOM is loaded 
+// and ready.
 for(let project of document.getElementsByClassName(PROJECT)) {
   project.addEventListener("click", openProject);
 }
+
+
+/**
+ * Fetches and displays a welcome message from DataServlet (at /data)
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:50]: Add a period at the end of your comments. Here and elsewhere.

Style guide says: End a complete sentence with a period, unless it's a question.
https://developers.google.com/style/periods

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: Ah, ok

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:29]: Oh that's good, will change.

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:35]: Fixed

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js]: Fixed
